src/geode/mesh/helpers/euclidean_distance_transform.cpp

```diff
@@ -0,0 +1,406 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/mesh/helpers/euclidean_distance_transform.h>
+
+#include <async++.h>
+
+#include <geode/basic/attribute_manager.h>
+#include <geode/basic/progress_logger.h>
+
+#include <geode/mesh/core/regular_grid_solid.h>
+#include <geode/mesh/core/regular_grid_surface.h>
+
+namespace geode
+{
+    template < index_t dimension >
+    class EuclideanDistanceTransform
+    {
+        using Index = GridCellIndices< dimension >;
+
+    public:
+        EuclideanDistanceTransform( const RegularGrid< dimension >& grid,
+            absl::Span< const GridCellIndices< dimension > > grid_cell_id,
+            absl::string_view distance_map_name )
+            : grid_{ grid }, squared_cell_length_{}, distance_map_{ nullptr }
+        {
+            for( const auto d : Range( dimension ) )
+            {
+                squared_cell_length_[d] = grid_.cell_length_in_direction( d )
+                                          * grid_.cell_length_in_direction( d );
+            }
+            distance_map_ =
+                grid_.cell_attribute_manager()
+                    .template find_or_create_attribute< VariableAttribute,
+                        double >( distance_map_name,
+                        std::numeric_limits< double >::max() );
+            for( const auto& cell_id : grid_cell_id )
+            {
+                distance_map_->set_value( grid_.cell_index( cell_id ), 0. );
+            }
+        }
+
+        std::shared_ptr< VariableAttribute< double > > distance_map() const
+        {
+            return distance_map_;
+        }
+
+        void compute_squared_distance_map();
+
+        void squared_root_filter()
+        {
+            async::parallel_for(
+                async::irange( index_t{ 0 }, grid_.nb_cells() ),
+                [this]( index_t cell ) {
+                    distance_map_->modify_value( cell, []( double& value ) {
+                        value = std::sqrt( value );
+                    } );
```

Comment:
This is just a side note: modify_value is required when you want to modify in place a value to avoid copying it. In the case of built-in types like double, int, etc it does not make a difference. So you can choose between the modify or the set.

Reply:
ok